### PR TITLE
PETSc BP3

### PR DESCRIPTION
**Pull request by jeremylt**
This is a first pass at a PETSc BP3 to clarify what PETSc solvers need.

### CODE REVIEW

File: examples/petsc/bp1.c [hunk 1]
```diff
@@ -97,44 +98,6 @@ static int CreateRestriction(Ceed ceed, const CeedInt melem[3],
   PetscFunctionReturn(0);
 }
 
-static int Setup(void *ctx, CeedInt Q,
```
Comment (jeremylt) [inline examples/petsc/bp1.c]: Stripped the QFunctions to a headder

File: examples/petsc/bp1.c [hunk 2]
```diff
@@ -376,21 +340,21 @@ int main(int argc, char **argv) {
   CeedQFunctionAddInput(qf_setup, "dx", 3, CEED_EVAL_GRAD);
   CeedQFunctionAddInput(qf_setup, "weight", 1, CEED_EVAL_WEIGHT);
   CeedQFunctionAddOutput(qf_setup, "rho", 1, CEED_EVAL_NONE);
-  CeedQFunctionAddOutput(qf_setup, "target", 1, CEED_EVAL_NONE);
```
Comment (jeremylt) [inline examples/petsc/bp1.c]: Renamed output "target" to "true_soln", added output "rhs"

File: examples/petsc/bp1.c [hunk 3]
```diff
@@ -446,14 +406,24 @@ int main(int argc, char **argv) {
   ierr = MatShellSetOperation(mat, MATOP_MULT, (void(*)(void))MatMult_Mass);
   CHKERRQ(ierr);
 
+  // Get RHS vector
   ierr = MatCreateVecs(mat, &rhs, NULL); CHKERRQ(ierr);
-  ierr = VecZeroEntries(X); CHKERRQ(ierr);
```
Comment (jeremylt) [inline examples/petsc/bp1.c]: Attempted to set up PETSc rhs vector, but something is wrong here

Comment (jeremylt) [reply examples/petsc/bp1.c]: Really, I think I set up/used the PETSc vectors and scatter wrong.

File: examples/petsc/bp1.c (hunk 2 above)
Comment (jeremylt) [reply examples/petsc/bp1.c]: I'd later like to change the vector to actually be called 'true_soln' as well, but I was going to wait until after we got everything running.

**Comment by jeremylt** [comment]
Update: The example compiles and runs, but I have something wrong still, as the error is too high.

**Comment by jeremylt** [comment]
To do:
Make BP3 actually work
Follow BP convention of 2 extra qpts
Use double dash for args

**Comment by jeremylt** [comment]
BP3 now converges but to the wrong answer

**Comment by jeremylt** [comment]
I picked the frequency and offset for the target function to zero the BCs, which is different than the MFEM version. I think that setting zero BCs does more closely focus on the libCEED part of the work though.

I think this branch is in good shape now. I want to make a couple of consistency changes, but then we can probably merge.

**Comment by jedbrown** [comment]
Do you think it makes sense to merge 164d7109079303b7ec02b71cafcc4350fe473cee (which implements Dirichlet boundary conditions, resulting in a nonsingular operator) into this branch? I kept all the `VecScatter`s in the interest of flexibility, but we could eliminate one if desired.

**Comment by jeremylt** [comment]
That commit seems reasonable to me. I think comes down to what we want to do with this example.

As an example, 164d710 better shows a basic way to incorporate boundary terms. And it is more in line with the MFEM BP3.

For benchmarks, this version more directly shows the performance of just libCEED, but I don't think we are there yet.

**Comment by jedbrown** [comment]
Did you mean to cite a different commit?

**Comment by jeremylt** [comment]
I meant to cite yours. I think your commit gives a better example of boundary conditions rather than leaving it somewhat ambiguously with zero Neumann boundary conditions.

**Pull request merged**